### PR TITLE
Composed ILS Driver

### PR DESCRIPTION
**Pull request by ThoWagen**
This new ILS Driver can be used to combine multiple other ILS Drivers. It allows to define multiple drivers for different driver methods and for many methods support drivers can be configured that provide additional information.

Here are some possible use cases for this:
1. Using the combination of PAIA and DAIA was implemented by letting the PAIA driver inherit from the DAIA driver. With this they can be split into two separate drivers and combined by this driver. PAIA can be used as the default main driver and DAIA can be configured for the methods getHolding, getStatus and getStatuses. 
2. This also allows to use PAIA in combination with other drivers for the record status methods.
3. One could deactivate all the methods that require a catalog login by using the NoILS-Driver as default main driver and only configure another driver for the methods that don't require a catalog login.
4. One can supplement the getHolding information with additional information, like summaries, notes etc. from an external API if they are not supported by the main driver.
5. One can get additional information about a user with the getMyProfile method from another source.

However, the driver does not support MultiILS for login. So currently one can only combine one driver that provides patronLogin with drivers that don't use patron information. An exception to this is when the other drivers can use the patron information of the driver that is used for the patronLogin. But this could be implemented in the future if it is required .

This driver was heavily inspired by the Multibackend driver. Much of the code both drivers have in common was moved into the new AbstractMultiDriver class.

To understand how the driver is supposed to work it might be a good idea to first look into the unit-tests for it.

### CODE REVIEW

**Review by demiankatz — COMMENTED** [review]
As the primary maintainer of MultiBackend, I imagine that this work may be of interest to @EreMaijala, so I've asked for a review from him (though I don't think there's a huge rush to complete it).

I also don't object to the name of CombinedDriver, but I wonder if something like ComposedDriver might be slightly more descriptive, since if I understand correctly, the purpose of this driver is to create a single interface by composing functionality from multiple places. Just a thought!

I'll give this a closer look myself after the 9.1 release process is wrapped up.

File: config/vufind/CombinedDriver.ini
Comment (EreMaijala) [inline config/vufind/CombinedDriver.ini]: Should this be `if *they* match`?

File: config/vufind/CombinedDriver.ini
Comment (EreMaijala) [inline config/vufind/CombinedDriver.ini]: Would it be imaginable that a different merge key would be needed for different support drivers, i.e. should it be specific to each support driver?

File: module/VuFind/src/VuFind/ILS/Driver/AbstractMultiDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/AbstractMultiDriver.php]: I think you'll need to carry over the copyrights for the existing moved code.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Default return value should be null.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: In Aleph driver this is `int|null`, so we should standardize on that.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Default return value can't be an array.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Default return value should be a string.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: This, like other methods that mention "a system message" in @return, should return an array such as `['success' => false, 'status' => 'ils_action_unavailable']` on failure.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
This is an interesting development that provides a nice layer of flexibility to the ILS drivers! I didn't see any major issues, so I've commented on some minor ones. 

I think that the default return values will need the most work. It might also make sense to make defaultCall throw an ILS exception by default unless instructed otherwise. This would eliminate the need to specify the default return values for cases where no suitable default is not available (or, make it more difficult to return an invalid value by accident).

**Comment by ThoWagen** [comment]
@demiankatz I renamed the driver to ComposedDriver.

@EreMaijala I think you are right about throwing exceptions instead of default return values. Since calling driver methods that are not supported should either not be possible or otherwise should be an error. I removed the default return values and now always an exception is thrown.

File: config/vufind/CombinedDriver.ini
Comment (ThoWagen) [reply config/vufind/CombinedDriver.ini]: Fixed

File: config/vufind/CombinedDriver.ini
Comment (ThoWagen) [reply config/vufind/CombinedDriver.ini]: We don't have a use case for this but I think it makes sense to have the option. I implemented it.

File: module/VuFind/src/VuFind/ILS/Driver/AbstractMultiDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/AbstractMultiDriver.php]: Fixed

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: I changed this here and also in the Demo driver.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Now, an exception is thrown if the method is not supported.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Now, an exception is thrown if the method is not supported.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Now, an exception is thrown if the method is not supported.

File: module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/CombinedDriver.php]: Now, an exception is thrown if the method is not supported.

**Comment by demiankatz** [comment]
Thanks, @ThoWagen -- I've renamed this PR to match the new name of the driver itself.

File: config/vufind/ComposedDriver.ini
Comment (EreMaijala) [inline config/vufind/ComposedDriver.ini]: I just noticed this one: I think we use `section` instead of `head`, so I'd change this for consistency.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
Looks good to me! Just one more nit to pick in a comment. :)

File: config/vufind/ComposedDriver.ini
Comment (ThoWagen) [reply config/vufind/ComposedDriver.ini]: done

**Review by EreMaijala — APPROVED** [review]

File: config/vufind/ComposedDriver.ini
Comment (demiankatz) [inline config/vufind/ComposedDriver.ini]: Typos -- should be "to be the same as"

File: config/vufind/ComposedDriver.ini
Comment (demiankatz) [inline config/vufind/ComposedDriver.ini]: "examples of how" might read a little better. "wont" should be "won't". I'm not entirely sure I understand why the Demo driver won't work, or why that is mentioned here (driver2 is not set to Demo in the example above).

File: module/VuFind/src/VuFind/ILS/Driver/ComposedDriver.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/ComposedDriver.php]: Since the configLoader and driverManager are dependencies of the base class, I think you could move this constructor there. I realize that MultiBackend has a different signature, but you could still move this constructor into the base class, and then simplify the MultiBackend constructor to call parent::__construct with the appropriate parameters.

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Demo.php:1533]: `?int` might be more concise here (though I suspect you copied this from Aleph, so if you prefer to keep it consistent with that, I have no objection).

File: module/VuFind/src/VuFind/ILS/Driver/ComposedDriverFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/ComposedDriverFactory.php]: If you decide to move the constructor, you could also rename this to AbstractMultiDriverFactory, so it would be reusable if a different subclass of the abstract class was created using the same signature.

File: module/VuFind/src/VuFind/ILS/Driver/MultiBackend.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/MultiBackend.php:1173]: Out of scope for this PR, but maybe something @EreMaijala should consider in a follow-up -- I see that $source can be false-y based on the code below, but the comment and type here don't seem to reflect that.

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/AbstractMultiDriverTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/AbstractMultiDriverTest.php]: This method is only relevant in MultiBackendTest (at least for now). Is it worth moving, or do you prefer to keep it in the abstract base class?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @ThoWagen and @EreMaijala! I gave this another look and found a few minor things to make suggestions about.

Once these are addressed, I think it's ready to merge, but I'm curious whether you think this should be included in release 9.1 or if it's safer to delay to 10.0 due to the significant refactoring of MultiBackend? I'm thinking that a 9.1 release is probably safe, since @EreMaijala is the primary user of MultiBackend and has approved this, but I'm happy to wait if either of you would prefer.

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/Demo.php:1533]: I did this since @EreMaijala suggested to standardize it here: https://github.com/vufind-org/vufind/pull/3112#discussion_r1335444249

It looks like it is done like this in a lot of places. So, I don't think this this violates any code style rules. But if you prefer `?int` I can change this.

File: config/vufind/ComposedDriver.ini
Comment (ThoWagen) [reply config/vufind/ComposedDriver.ini]: fixed

File: config/vufind/ComposedDriver.ini
Comment (ThoWagen) [reply config/vufind/ComposedDriver.ini]: fixed. I also explained why the Demo driver won't work.

File: module/VuFind/src/VuFind/ILS/Driver/ComposedDriver.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/ComposedDriver.php]: done

File: module/VuFind/src/VuFind/ILS/Driver/ComposedDriverFactory.php
Comment (ThoWagen) [reply module/VuFind/src/VuFind/ILS/Driver/ComposedDriverFactory.php]: done

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/AbstractMultiDriverTest.php
Comment (ThoWagen) [reply module/VuFind/tests/unit-tests/src/VuFindTest/ILS/Driver/AbstractMultiDriverTest.php]: done

**Comment by ThoWagen** [comment]
It is absolutely fine for me if it is included in 10.0

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Demo.php:1533]: Feel free to leave it alone for now -- since it's consistent with other things, there's no need to change it. I think in the long term I'd prefer the `?` syntax instead of the `|null` syntax since it is more concise -- but fixing that should probably become a separate project, and it's best to back it up with some kind of style checker if we make that decision. All out of scope for what you're doing here. :-)

**Review by demiankatz — APPROVED** [review]
Thanks, @ThoWagen -- this looks good to me. I've just made one small edit to add a missing word in a comment.

If @EreMaijala wants to include this in release 9.1, he's welcome to hit the merge button. Otherwise, I'll merge this as soon as we're done with work on release 9.1.

**Comment by demiankatz** [comment]
(Setting a 10.0 milestone to be sure this doesn't get forgotten if it isn't merged sooner).

File: module/VuFind/src/VuFind/ILS/Driver/Demo.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Driver/Demo.php:1533]: I didn't mean to imply that you should use `int|null`as `?int` is indeed better, though I also value keeping the docs of same function consistent. Perhaps we could also add more interfaces and use `@inheritDoc` to avoid copy-pasting and ensure that the method definitions don't diverge.

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Driver/Demo.php:1533]: The problem with inheritDoc is that php codesniffer doesn't support it properly, so we'd either need to disable comment validation or ignore lots of errors. I wish this would get addressed, but I think it's a very hard problem to solve based on the way codesniffer works.

**Pull request merged**
